Middleware refactor

danielabbasi opened the pull request:
- Split out middleware into 3 different ones: initialStartupMiddleware, search, and chart and put them into their corresponding feature files. 

- Created a socketService which follows a singleton design pattern to allow for coordinated access to the socket, throughout the entire application. Ensures that it’s only instantiated once, and provides easy access to the single instance.

- Created a event constants file for the web socket events and am using it throughout client side.

Sharakai commented on stock-tracker/src/features/chart/redux/chartMiddleware.js:
```diff
@@ -0,0 +1,18 @@
+export const chartMiddleware = ({
+  socketService
+}) => store => next => action => {
+  const result = next(action);
```
Why do you call `next(action)` at the start, and not at the end like `return next(action)`

Same with all your other middleware.

danielabbasi replied:
Have had a look and changed it so I am now using the payload to emit rather than getting from store

Sharakai replied:
Eh? I don't get it...

Sharakai commented on stock-tracker/src/features/search/redux/searchMiddleware.js:
```diff
@@ -0,0 +1,40 @@
+    socketService
+      .create()
+      .emit(SEARCH_INPUT, store.getState().search.searchInput);
+    socketService.create().on(SUGGESTIONS, suggestions => {
```
**Memory leak**: Why're you creating a new `on(SUGGESTIONS, ...` handler here? This should be attached once to the socket, whereas it's currently being attached on every `ADD_SEARCH_INPUT` action

Sharakai replied:
And why is it not using your `socketService.createSocketSubscription` function?

danielabbasi replied:
I thought it just receives the already created socket, have now attached it to a socket variable and reusing that variable.

Sharakai replied:
This hasn't fixed the memory leak. I wasn't referring to the use of the `socket` variable - That isn't a memory leak.
I'm referring to the fact that you're creating a new `.on` handler on every `ADD_SEARCH_INPUT` action, and never calling `.off` to remove it.

danielabbasi replied:
My bad. I don't need this within the search input. Realized the subscription has already been created within subscriptions.js on start up.

Sharakai commented on stock-tracker/src/store/initialStartupMiddleware.js:
```diff
@@ -0,0 +1,13 @@
+  socketService
+}) => store => next => action => {
+  if (action.type === INITIAL_STARTUP) {
+    console.info("Application has started ");
```
nit: redundant space at the end

Sharakai commented on stock-tracker/src/store/store.js:
```diff
@@ -0,0 +1,20 @@
+import { chartMiddleware } from "../features/chart/redux/chartMiddleware";
+import { socketService } from "../socket";
+
+const middlwares = [
```
sp: `middleware`. It's a singulare tantum.

Sharakai replied:
You're still missing the `e`

danielabbasi replied:
oops

Sharakai commented on stock-tracker/src/store/subscriptions.js:
```diff
@@ -0,0 +1,60 @@
+import {
+  setChartDataAction,
+  setChartErrorAction
+} from "../features/chart/redux/actions";
```
Note: These imports show the nicety of having an `index.js` file in the feature root, which defines an "API" of the folder and exports the component, reducer, middleware, and externally required actions.
The import would then just be `"../features/chart";`

Sharakai commented on stock-tracker/src/store/subscriptions.js:
```diff
@@ -0,0 +1,60 @@
+
+export const createSocketSubscriptions = dispatch => {
+  socketService.create();
+  const unsubscribeFns = [
```
This array of `[action type, action]` tuples could be extracted out of this function into a constant above.
You could also change it to a map, if you wanted to make sure each action type is unique and handled only once, but that's preference.

danielabbasi commented:
Sorry for big push. Added index.js to each component and changed imports. Also made the changes concerning next(action) in each middleware and solved memory leak issue.

Sharakai commented on stock-tracker/src/socket/eventActions.js:
```diff
@@ -0,0 +1,45 @@
+import {
```
Pref: I would _personally_ group all this into a namespaced import:
```
import * as EventType from "./eventTypes";
```
and then use them like
```
[EventType.STOCK_DATA, setResponseAction]
```

Sharakai commented on stock-tracker/src/features/search/redux/searchMiddleware.js:
```diff
@@ -0,0 +1,32 @@
+export const searchMiddleware = ({
+  socketService
+}) => store => next => action => {
+  const socket = socketService.create();
```
This will actually end up creating the WebSocket on the Redux `@@INIT` event, which may not be what you want.
I recommend moving this inside of the `if` blocks

Sharakai commented on stock-tracker/package.json:
```diff
@@ -41,7 +41,7 @@
   "devDependencies": {
     "husky": "^3.0.8",
-    "nodemon": "^1.19.2",
+    "nodemon": "^1.19.4",
```
This change has no relevant to the rest of the PR, and shouldn't be here.
I'll accept leaving it in for now, but try to remember not to bleed into other features / changes.
